src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php

```diff
@@ -41,42 +41,61 @@ trait PriorityTaggedServiceTrait
             $needsIndexes = $tagName->needsIndexes();
+            $defaultPriorityMethod = $tagName->getDefaultPriorityMethod();
             $tagName = $tagName->getTag();
         }
 
         $services = [];
 
         foreach ($container->findTaggedServiceIds($tagName, true) as $serviceId => $attributes) {
-            $priority = isset($attributes[0]['priority']) ? $attributes[0]['priority'] : 0;
+            $class = $container->getDefinition($serviceId)->getClass();
+            $class = $container->getParameterBag()->resolveValue($class) ?: null;
+            $reflectionClass = $container->getReflectionClass($class);
+
+            $priority = 0;
+            if (null !== $defaultPriorityMethod && null !== $reflectionClass && $reflectionClass->hasMethod($defaultPriorityMethod)) {
+                if (!($priorityReflMethod = $reflectionClass->getMethod($defaultPriorityMethod))->isStatic()) {
+                    throw new InvalidArgumentException(sprintf('Default priority method "%s::%s()" of the "%s"-tagged collection on service "%s" must be static.', $class, $defaultPriorityMethod, $tagName, $serviceId));
+                }
+
+                if (!$priorityReflMethod->isPublic()) {
+                    throw new InvalidArgumentException(sprintf('Default priority method "%s::%s()" of the "%s"-tagged collection on service "%s" should be public.', $class, $defaultPriorityMethod, $tagName, $serviceId));
+                }
+
+                $priority = $priorityReflMethod->invoke(null);
+
+                if (!\is_int($priority)) {
+                    throw new InvalidArgumentException(sprintf('Default priority method "%s::%s()" of the "%s"-tagged collection on service "%s" should return an integer, got %s.', $class, $defaultPriorityMethod, $tagName, $serviceId, \gettype($priority)));
+                }
+            }
+
+            $priority = $attributes[0]['priority'] ?? $priority;
 
             if (null === $indexAttribute && !$needsIndexes) {
                 $services[$priority][] = new Reference($serviceId);
 
                 continue;
             }
 
-            $class = $container->getDefinition($serviceId)->getClass();
-            $class = $container->getParameterBag()->resolveValue($class) ?: null;
-
             if (null !== $indexAttribute && isset($attributes[0][$indexAttribute])) {
                 $services[$priority][$attributes[0][$indexAttribute]] = new TypedReference($serviceId, $class, ContainerBuilder::EXCEPTION_ON_INVALID_REFERENCE, $attributes[0][$indexAttribute]);
 
                 continue;
             }
 
-            if (!$r = $container->getReflectionClass($class)) {
+            if (null === $reflectionClass) {
                 throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $serviceId));
             }
 
-            $class = $r->name;
+            $class = $reflectionClass->name;
```

Comment:
let's keep `$r` as the name here, that will reduce the diff, thus reduce future merge conflicts